### PR TITLE
#40839445 - Add ability to Gift NFT

### PR DESCRIPTION
**Pull request by VitaliyVK777**
The onwer of an NFT should have the ability to give a copy of an NFT “for free” to a specific user and just pay a small transfer fee 10 PSL. The user would need to “accept” the gifted copies and should pay a small transfer fee. The reason this makes sense is because some creators will want to give away NFTs to their fans instead of selling the NFT. And you can’t really implement this as “selling the NFT copy for a price of 1 PSL because then the “seller” can’t choose the “buyer”— someone else could jump in quickly and get the NFT for 1 PSL.

To implement this feature the owner of NFT should set the recipient pastel ID of an NFT gift. We use the "recipientPastelID" instead of "reserved" field in the Sell ticket for this purpose to have the compatibility with the old clients. Selling of an NFT copy is allowed only to this "recipientPastelID" if it is not empty.

So the owner of an NFT gift can create now the Sell ticket with price "0" and the "recipientPastelID" to gift an NFT copy. Then the "recipientPastelID" is checked in the "Buy" and "Trade" tickets. The "recipientPastelID" parameter has been added to the "tickets register sell" RPC command.

Also additionally the user can sell an NFT copy directly to the "recipientPastelID" if he wants. In this case the owner of the NFT copy can create now the Sell ticket with some price and the "recipientPastelID" parameter.

Tests completed: 7
successes 7; failures: 0

### CODE REVIEW

File: src/mnode/mnode-pastel.cpp [hunk 1]
```diff
@@ -792,368 +857,395 @@ std::vector<CArtActivateTicket> CArtActivateTicket::FindAllTicketByArtistHeight(
...
-    std::stringstream ss;
-    ss << pastelID;
-    ss << artTnxId;
-    ss << askedPrice;
-    ss << copyNumber;
-    ss << activeBefore;
-    ss << activeAfter;
-    ss << m_nTimestamp;
-    return ss.str();
+std::string CArtSellTicket::ToStr() const noexcept {
```
Comment (vadymeng0) [inline src/mnode/mnode-pastel.cpp:888]: Lots of the changes is about code-formatting differences only.
This makes the changes hard to review.
Does the coding convention changed intentionally?

**Review by a-ok123 — CHANGES_REQUESTED** [review]
There are too many changes - this SHALL NOT be merged until further notice

File: src/mnode/mnode-pastel.cpp [hunk 2]
```diff
@@ -603,14 +603,79 @@ bool common_validation(const T& ticket, bool preReg, const std::string& strTnxId
...
-    std::string err;
     if (!pastelTicket->IsValid(false, ++depth)) {
-      throw std::runtime_error(strprintf("The %s ticket with this txid [%s] is invalid - %s", prevTicket, strTnxId, err));
+      throw std::runtime_error(strprintf("The %s ticket with this txid [%s] is invalid", prevTicket, strTnxId));
     }
     
     return true;
 }
 
+bool trade_copy_validation(const std::string& nftTnxId, const std::vector<unsigned char>& signature) {
```
[the remark this answers is not in the document]
Comment (VitaliyVK777) [reply src/mnode/mnode-pastel.cpp:613]: ok

File: src/mnode/mnode-pastel.cpp [hunk 3]
```diff
@@ -603,14 +603,79 @@ bool common_validation(const T& ticket, bool preReg, const std::string& strTnxId
...
+  if (masterNodeCtrl.masternodeSync.IsSynced()) {
```
[the remark this answers is not in the document]
Comment (VitaliyVK777) [reply src/mnode/mnode-pastel.cpp:614]: ok

**Pull request closed without merging**